[Bug] AnkiDroid doesn't open: FragmentManager has not been attached to a host

```
2021-04-14 18:44:34.774 21612-21612/com.ichi2.anki D/ApplicationLoaders: Returning zygote-cached class loader: /system/framework/android.test.base.jar
2021-04-14 18:44:35.288 21612-21612/com.ichi2.anki D/NetworkSecurityConfig: Using Network Security Config from resource network_security_config debugBuild: true
2021-04-14 18:44:35.290 21612-21612/com.ichi2.anki D/NetworkSecurityConfig: Using Network Security Config from resource network_security_config debugBuild: true
2021-04-14 18:44:35.354 21612-21612/com.ichi2.anki I/ACRA: ACRA is disabled for com.ichi2.anki, initializing...
2021-04-14 18:44:35.389 21612-21612/com.ichi2.anki D/AnkiDroid: Startup - Application Start
2021-04-14 18:44:35.391 21612-21612/com.ichi2.anki I/UsageAnalytics: initialize()
2021-04-14 18:44:35.392 21612-21612/com.ichi2.anki D/UsageAnalytics: App tracking id 'tid' = UA-125800786-1
2021-04-14 18:44:35.428 21612-21612/com.ichi2.anki I/WebViewFactory: Loading com.google.android.webview version 89.0.4389.105 (code 438910533)
2021-04-14 18:44:35.456 21612-21612/com.ichi2.anki W/com.ichi2.anki: Accessing hidden method Landroid/os/Trace;->isTagEnabled(J)Z (greylist, reflection, allowed)
2021-04-14 18:44:35.456 21612-21612/com.ichi2.anki W/com.ichi2.anki: Accessing hidden method Landroid/os/Trace;->traceBegin(JLjava/lang/String;)V (greylist, reflection, allowed)
2021-04-14 18:44:35.456 21612-21612/com.ichi2.anki W/com.ichi2.anki: Accessing hidden method Landroid/os/Trace;->traceEnd(J)V (greylist, reflection, allowed)
2021-04-14 18:44:35.456 21612-21612/com.ichi2.anki W/com.ichi2.anki: Accessing hidden method Landroid/os/Trace;->asyncTraceBegin(JLjava/lang/String;I)V (greylist, reflection, allowed)
2021-04-14 18:44:35.456 21612-21612/com.ichi2.anki W/com.ichi2.anki: Accessing hidden method Landroid/os/Trace;->asyncTraceEnd(JLjava/lang/String;I)V (greylist, reflection, allowed)
2021-04-14 18:44:35.473 21612-21612/com.ichi2.anki I/cr_LibraryLoader: Loaded native library version number "89.0.4389.105"
2021-04-14 18:44:35.474 21612-21612/com.ichi2.anki I/cr_CachingUmaRecorder: Flushed 4 samples from 4 histograms.
2021-04-14 18:44:35.477 21612-21612/com.ichi2.anki I/TetheringManager: registerTetheringEventCallback:com.ichi2.anki
2021-04-14 18:44:35.653 21612-21612/com.ichi2.anki I/GoogleAnalyticsImpl: Not participating in analytics sample (sample percentage vs random: 10 32)
2021-04-14 18:44:35.653 21612-21612/com.ichi2.anki I/UsageAnalytics: setOptIn(): from false to false
2021-04-14 18:44:35.653 21612-21612/com.ichi2.anki I/GoogleAnalyticsImpl: Not participating in analytics sample (sample percentage vs random: 10 92)
2021-04-14 18:44:35.654 21612-21612/com.ichi2.anki D/UsageAnalytics: setOptIn() optIn / sAnalytics.config().enabled(): false/false
2021-04-14 18:44:35.654 21612-21612/com.ichi2.anki I/UsageAnalytics: setDryRun(): true, warning dryRun is experimental
2021-04-14 18:44:35.663 21612-21612/com.ichi2.anki I/CompatV26: Creating notification channel with id/name: General Notifications/AnkiDroid
2021-04-14 18:44:35.664 21612-21612/com.ichi2.anki I/CompatV26: Creating notification channel with id/name: Synchronization/Synchronization
2021-04-14 18:44:35.666 21612-21612/com.ichi2.anki I/CompatV26: Creating notification channel with id/name: Global Reminders/Cards due
2021-04-14 18:44:35.667 21612-21612/com.ichi2.anki I/CompatV26: Creating notification channel with id/name: Deck Reminders/Reminders
2021-04-14 18:44:35.702 21612-21612/com.ichi2.anki I/AnkiDroidApp: AnkiDroidApp: Starting Services
2021-04-14 18:44:35.706 21612-21612/com.ichi2.anki I/CollectionHelper: Begin openCollection: /storage/emulated/0/AnkiDroidA/collection.anki2
2021-04-14 18:44:35.733 21612-21612/com.ichi2.anki D/RustSupportSQLiteOpenHelper: createRustSupportSQLiteDatabase
2021-04-14 18:44:36.012 21612-21612/com.ichi2.anki I/Collection: Opening Collection Log
2021-04-14 18:44:36.016 21612-21612/com.ichi2.anki D/Collection: [1618422276] Collection.java:<init>(): /storage/emulated/0/AnkiDroidA/collection.anki2,  2.15alpha41
2021-04-14 18:44:36.075 21612-21612/com.ichi2.anki I/NativeMethods: ExecuteCommand: localMinutesWest
2021-04-14 18:44:36.084 21612-21612/com.ichi2.anki D/Collection: [1618422276] SchedV2.java:_updateCutoff(): 32,  1618455600
2021-04-14 18:44:36.085 21612-21612/com.ichi2.anki I/NativeMethods: ExecuteCommand: localMinutesWest
2021-04-14 18:44:36.085 21612-21612/com.ichi2.anki I/CollectionHelper: End openCollection: /storage/emulated/0/AnkiDroidA/collection.anki2
2021-04-14 18:44:36.086 21612-21612/com.ichi2.anki I/BootService: Executing Boot Service
2021-04-14 18:44:36.110 21612-21612/com.ichi2.anki D/IntentHandler: onCreate()
2021-04-14 18:44:36.122 21612-21612/com.ichi2.anki D/IntentHandler: onCreate() performing default action
2021-04-14 18:44:36.123 21612-21612/com.ichi2.anki I/IntentHandler: Launching DeckPicker
2021-04-14 18:44:36.180 21612-21612/com.ichi2.anki D/AnkiDroidApp: AnkiDroidApp::getLanguageConfig - setting locale to en_GB
2021-04-14 18:44:36.208 21612-21612/com.ichi2.anki D/DeckPicker: onCreate()
2021-04-14 18:44:36.210 21612-21612/com.ichi2.anki I/DeckPicker: AnkiDroid is being updated and a collection already exists.
2021-04-14 18:44:36.217 21612-21612/com.ichi2.anki D/AndroidRuntime: Shutting down VM
    
    
    --------- beginning of crash
2021-04-14 18:44:36.219 21612-21612/com.ichi2.anki E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.ichi2.anki, PID: 21612
    java.lang.RuntimeException: Unable to start activity ComponentInfo{com.ichi2.anki/com.ichi2.anki.DeckPicker}: java.lang.IllegalStateException: FragmentManager has not been attached to a host.
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3431)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3595)
        at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:85)
        at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2066)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:223)
        at android.app.ActivityThread.main(ActivityThread.java:7660)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
     Caused by: java.lang.IllegalStateException: FragmentManager has not been attached to a host.
        at androidx.fragment.app.FragmentManager.enqueueAction(FragmentManager.java:1880)
        at androidx.fragment.app.BackStackRecord.commitInternal(BackStackRecord.java:329)
        at androidx.fragment.app.BackStackRecord.commit(BackStackRecord.java:294)
        at androidx.fragment.app.DialogFragment.show(DialogFragment.java:277)
        at com.ichi2.anki.AnkiActivity.showDialogFragment(AnkiActivity.java:444)
        at com.ichi2.anki.AnkiActivity.showDialogFragment(AnkiActivity.java:429)
        at com.ichi2.anki.DeckPicker.showStartupScreensAndDialogs(DeckPicker.java:1191)
        at com.ichi2.anki.DeckPicker.handleStartup(DeckPicker.java:541)
        at com.ichi2.anki.DeckPicker.onCreate(DeckPicker.java:456)
        at android.app.Activity.performCreate(Activity.java:8000)
        at android.app.Activity.performCreate(Activity.java:7984)
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1309)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3404)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3595) 
        at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:85) 
        at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135) 
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2066) 
        at android.os.Handler.dispatchMessage(Handler.java:106) 
        at android.os.Looper.loop(Looper.java:223) 
        at android.app.ActivityThread.main(ActivityThread.java:7660) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947) 
2021-04-14 18:44:36.219 21612-21612/com.ichi2.anki D/UsageAnalytics: sendAnalyticsException() description/fatal: java.lang.IllegalStateException: FragmentManager has not been attached to a host./true
2021-04-14 18:44:36.224 21612-21612/com.ichi2.anki I/ACRA: ACRA is disabled for com.ichi2.anki - forwarding uncaught Exception on to default ExceptionHandler
2021-04-14 18:44:36.232 21612-21612/com.ichi2.anki I/Process: Sending signal. PID: 21612 SIG: 9
```

--- a/AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java
@@ -452,10 +452,10 @@ public class DeckPicker extends NavigationDrawerActivity implements
 
         //we need to restore here, as we need it before super.onCreate() is called.
         restoreWelcomeMessage(savedInstanceState);
-        handleStartup();
 
         // Then set theme and content view
         super.onCreate(savedInstanceState);
+        handleStartup();
         setContentView(R.layout.homescreen);
         View mainView = findViewById(android.R.id.content);
 

--- a/AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java
@@ -1187,7 +1187,7 @@ public class DeckPicker extends NavigationDrawerActivity implements
             Timber.i("AnkiDroid is being updated and a collection already exists.");
             // The user might appreciate us now, see if they will help us get better?
             if (!preferences.contains(UsageAnalytics.ANALYTICS_OPTIN_KEY)) {
-                showDialogFragment(DeckPickerAnalyticsOptInDialog.newInstance());
+                displayAnalyticsOptInDialog();
             }
 
             // For upgrades, we check if we are upgrading
@@ -1324,6 +1324,12 @@ public class DeckPicker extends NavigationDrawerActivity implements
             onFinishedStartup();
         }
     }
+
+    @VisibleForTesting
+    protected void displayAnalyticsOptInDialog() {
+        showDialogFragment(DeckPickerAnalyticsOptInDialog.newInstance());
+    }
+
 
     protected long getPreviousVersion(SharedPreferences preferences, long current) {
         long previous;

--- a/AnkiDroid/src/test/java/com/ichi2/anki/DeckPickerTest.java
+++ b/AnkiDroid/src/test/java/com/ichi2/anki/DeckPickerTest.java
@@ -13,7 +13,6 @@ import com.ichi2.libanki.sched.AbstractSched;
 import com.ichi2.testutils.BackendEmulatingOpenConflict;
 import com.ichi2.testutils.BackupManagerTestUtilities;
 
-import org.junit.Ignore;
 import org.junit.Test;
 import org.junit.runner.RunWith;
 import org.robolectric.Robolectric;
@@ -249,7 +248,6 @@ public class DeckPickerTest extends RobolectricTest {
     }
 
     @Test
-    @Ignore("8610")
     public void deckPickerOpensWithHelpMakeAnkiDroidBetterDialog() {
         // Refactor: It would be much better to use a spy - see if we can get this into Robolecteic
         try {

--- a/AnkiDroid/src/test/java/com/ichi2/anki/DeckPickerTest.java
+++ b/AnkiDroid/src/test/java/com/ichi2/anki/DeckPickerTest.java
@@ -11,7 +11,9 @@ import com.ichi2.libanki.Collection;
 import com.ichi2.libanki.DeckConfig;
 import com.ichi2.libanki.sched.AbstractSched;
 import com.ichi2.testutils.BackendEmulatingOpenConflict;
+import com.ichi2.testutils.BackupManagerTestUtilities;
 
+import org.junit.Ignore;
 import org.junit.Test;
 import org.junit.runner.RunWith;
 import org.robolectric.Robolectric;
@@ -246,8 +248,30 @@ public class DeckPickerTest extends RobolectricTest {
         }
     }
 
+    @Test
+    @Ignore("8610")
+    public void deckPickerOpensWithHelpMakeAnkiDroidBetterDialog() {
+        // Refactor: It would be much better to use a spy - see if we can get this into Robolecteic
+        try {
+            InitialActivityTest.grantWritePermissions();
+            BackupManagerTestUtilities.setupSpaceForBackup(getTargetContext());
+            // We don't show it if the user is new.
+            AnkiDroidApp.getSharedPrefs(getTargetContext()).edit().putString("lastVersion", "0.1").apply();
+
+            DeckPickerEx d = super.startActivityNormallyOpenCollectionWithIntent(DeckPickerEx.class, new Intent());
+
+            assertThat("Analytics opt-in should be displayed", d.mDisplayedAnalyticsOptIn, is(true));
+
+        } finally {
+            InitialActivityTest.revokeWritePermissions();
+            BackupManagerTestUtilities.reset();
+        }
+    }
+
     private static class DeckPickerEx extends DeckPicker {
         private int mDatabaseErrorDialog;
+        private boolean mDisplayedAnalyticsOptIn;
+
 
         @Override
         public void showDatabaseErrorDialog(int id) {
@@ -256,6 +280,13 @@ public class DeckPickerTest extends RobolectricTest {
 
         public void onStoragePermissionGranted() {
             onRequestPermissionsResult(DeckPicker.REQUEST_STORAGE_PERMISSION, new String[] { "" }, new int[] { PackageManager.PERMISSION_GRANTED });
+        }
+
+
+        @Override
+        protected void displayAnalyticsOptInDialog() {
+            this.mDisplayedAnalyticsOptIn = true;
+            super.displayAnalyticsOptInDialog();
         }
     }
 }

--- a/AnkiDroid/src/test/java/com/ichi2/anki/InitialActivityTest.java
+++ b/AnkiDroid/src/test/java/com/ichi2/anki/InitialActivityTest.java
@@ -67,14 +67,25 @@ public class InitialActivityTest extends RobolectricTest {
     }
 
     public static void setupForDatabaseConflict() {
-        ShadowApplication app = Shadows.shadowOf((Application) ApplicationProvider.getApplicationContext());
-        app.grantPermissions(android.Manifest.permission.WRITE_EXTERNAL_STORAGE, android.Manifest.permission.READ_EXTERNAL_STORAGE);
+        grantWritePermissions();
         ShadowEnvironment.setExternalStorageState("mounted");
     }
 
+
+
     public static void setupForDefault() {
+        revokeWritePermissions();
+        ShadowEnvironment.setExternalStorageState("removed");
+    }
+
+
+    protected static void grantWritePermissions() {
+        ShadowApplication app = Shadows.shadowOf((Application) ApplicationProvider.getApplicationContext());
+        app.grantPermissions(android.Manifest.permission.WRITE_EXTERNAL_STORAGE, android.Manifest.permission.READ_EXTERNAL_STORAGE);
+    }
+
+    protected static void revokeWritePermissions() {
         ShadowApplication app = Shadows.shadowOf((Application) ApplicationProvider.getApplicationContext());
         app.denyPermissions(android.Manifest.permission.WRITE_EXTERNAL_STORAGE, android.Manifest.permission.READ_EXTERNAL_STORAGE);
-        ShadowEnvironment.setExternalStorageState("removed");
     }
 }

